src/smk/RenderTarget.cpp

```diff
@@ -145,13 +145,13 @@ void RenderTarget::SetShaderProgram(ShaderProgram* shader_program) {
 
 /// @brief Return the default predefined 2D shader program. It is bound by
 /// default.
-ShaderProgram& RenderTarget::shader_program_2d() const {
-  return *shader_program_2d_.get();
+ShaderProgram* RenderTarget::shader_program_2d() const {
+  return shader_program_2d_.get();
```

Comment:
Why choose to return unsafe pointer here?

Reply:
A pointer and a reference is the same thing. Why one is unsafe and not the other?

Reply:
That's not true. There's many reasons why google C++ standards say to use smart pointers and ownership wrappers because pointers are inherently unsafe. Because the API is intended to return something that should exist within the duration of the entire API, it's also a hint to the end user to use a reference to show them that they will be using the _reference_ to something.

Reply:
Any updates on this? I'm waiting to resolve this feature before porting a few libraries I wrote for SFML to be used with smk :D

Reply:
> That's not true.

I don't believe returning a reference means the object will be available during the entire API.

A reference is just a sugar syntax over pointers.

I prefer pointer here, because the API:
```
Window::SetShaderProgram(ShaderProgram*)
```
use a pointer.
I don't want to write:
```
window->SetShaderProgram(*(window->shader_program_3d()))
```
instead of:
```
window->SetShaderProgram(window->shader_program_3d())
```

I am not very proud the RenderTarget hold the default shaders, but I haven't really found a better way for now.

Reply:
reference is more than sugar syntax, it is a type identifier. A pointer is expected to change what it _points to_ and is expected to have a temporary lifetime. A reference on the otherhand is never going to change what it references and while the lifetime could be bunk, in this case we are returning a build-in shader program that will exist during the entire lifetime of the API. Returning a pointer that we can mutate is dangerous. A pointer return value is also an indicator that it could return nullptr. 

Even if you do not agree with these points, the google standard also states to not use raw pointers.

> window->SetShaderProgram(*(window->shader_program_3d()))
This should actually be 
`window->SetShaderProgram(&window->shader_program_3d())`

And you could accept a `nullptr` as a shortcut to check for it and then switch to the internal shader program

```
RenderTarget::SetShaderProgram(Shader* shader) {
    if(shader == nullptr) {
        shader = shader_program_3d_; // switch to default shader program
    }
}
```

Or to be explicit, we could provide a `UseShaderProgramDefault2D()` or the like.

> I am not very proud the RenderTarget hold the default shaders

The only optimization is to make them static. I made SFML GL ES 2 compliant and had to also keep a default shader program in the SFML API internals. I don't think there is a better way to do this.

Reply:
> Even if you do not agree with these points, the google standard also states to not use raw pointers.

I don't think so. It [says](https://google.github.io/styleguide/cppguide.html#Ownership_and_Smart_Pointers):
> If dynamic allocation is necessary, prefer to keep ownership with the code that allocated it. If other code needs access to the object, consider passing it a copy, or passing a pointer or reference without transferring ownership. Prefer to use std::unique_ptr to make ownership transfer explicit. For example:

However, you are starting to convince me. I think I will switch to reference. However it means I would have to provide SetShaderProgram two version one about pointer and one about reference.

Reply:
> Prefer to use std::unique_ptr to make ownership transfer explicit.

Or at least this point.

Though I'd also argue against providing 2 overloads for the same function. Either take a reference, pointer, but not both. Though I'm starting to think a shared ptr makes more sense. If the shader built goes out of scope, it may crash the entire program because it is deleted. A shared ptr would track uses and keep working until the render target is explicitly instructed to use another shader.

Reply:
It is not entirely clear to me who own the shader. It is more or less bound to a Window, since this is owned by the GPU and the OpenGL context is owned by the Window.

So maybe we should only make the Window able to create Shader. On the other side, I have in my games several function that are drawing inside a RenderTarget which is sometimes a FrameBuffer, sometimes a Window. I am not sure what is best here.